### ISSUE TITLE
In a paper, the old model and the model being trained are supposed to be at different time steps. Why are they the same in `grpo_trainer`?

### ISSUE THREAD

**Issue by YooSungHyun** [huggingface/trl#3294]
### Reproduction

<img width="1127" alt="Image" src="https://github.com/user-attachments/assets/8663953f-edc3-4280-9a2e-e7d9604db384" />

if step == 0
old_model == model

if step == 1
old_model: step 0 model
model: step 1 model (1 back-prob)

if step == 2
old_model: step 1 model
model: step 2 model (2 back-prob)

but, grpo_trainer is always same between old_model and model

https://github.com/huggingface/trl/blob/main/trl/trainer/grpo_trainer.py#L844
```python
with torch.no_grad():
    # When using num_iterations == 1, old_per_token_logps == per_token_logps, so we can skip it's
    # computation here, and use per_token_logps.detach() instead.
    if self.num_iterations > 1:
        old_per_token_logps = self._get_per_token_logps(
            **self.model**, prompt_completion_ids, attention_mask, logits_to_keep
        )
    else:
        old_per_token_logps = None
```
https://github.com/huggingface/trl/blob/main/trl/trainer/grpo_trainer.py#L1034
```python
@profiling_decorator
def compute_loss(self, **model**, inputs, return_outputs=False, num_items_in_batch=None):
    if return_outputs:
        raise ValueError("The GRPOTrainer does not support returning outputs")
    if self.use_liger_loss:
        # Compute the loss using the liger grpo loss
        return self.compute_liger_loss(model, inputs)
    else:
        return self._compute_loss(model, inputs)

def _compute_loss(self, **model**, inputs):
    # Compute the per-token log probabilities for the model
    prompt_ids, prompt_mask = inputs["prompt_ids"], inputs["prompt_mask"]
    completion_ids, completion_mask = inputs["completion_ids"], inputs["completion_mask"]
    input_ids = torch.cat([prompt_ids, completion_ids], dim=1)
    attention_mask = torch.cat([prompt_mask, completion_mask], dim=1)
    logits_to_keep = completion_ids.size(1)  # we only need to compute the logits for the completion tokens

    per_token_logps = self._get_per_token_logps(**model**, input_ids, attention_mask, logits_to_keep)
```
this two model is just self.model right???
In this case, since the ratio between the model and old_model becomes 1, doesn’t it effectively reduce to standard PPO with G-sampled rewards and a separate KL penalty term?

In my opinion, the _prepare_input function should consider grad_accumulation_step and periodically deepcopy the old_model. Also, in the generate_and_score part—specifically when computing log probabilities—it should use the old_model instead. What do you think?

### System Info

- Platform: Linux-5.15.0-60-generic-x86_64-with-glibc2.35
- Python version: 3.11.12
- TRL version: 0.17.0.dev0
- PyTorch version: 2.5.1
- CUDA device(s): NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB
- Transformers version: 4.51.2
- Accelerate version: 1.6.0
- Accelerate config: not found
- Datasets version: 3.5.0
- HF Hub version: 0.30.2
- bitsandbytes version: 0.45.5
- DeepSpeed version: 0.16.5
- Diffusers version: not installed
- Liger-Kernel version: 0.5.5
- LLM-Blender version: not installed
- OpenAI version: 1.72.0
- PEFT version: 0.15.1
- vLLM version: not installed

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by LeonEricsson** [comment]
At each step $\pi_{\theta_{old}}$ and $\pi_{\theta}$ are the same up until line 10 when the GRPO iterations start updating $\pi_{\theta}$ away from $\pi_{\theta_{old}}$. If you are using `num_iterations=1` in the config then you are correct that the ratio is always 1.

**Comment by LeonEricsson** [comment]
Yes, you are misunderstanding the implementation. Let me attempt to clarify this further with an example:

Assume `num_iterations=2`. I'll walk through the process step by step:

1. At the beginning of each step (referencing Algorithm 1), we generate samples and completions, then score these completions to compute the loss and update the policy ($\pi_{\theta}$). This is orchestrated from

https://github.com/huggingface/trl/blob/d625c5533a6b1c84d3565c8080857f6bb81c538a/trl/trainer/grpo_trainer.py#L735-L750

2. In the first step (`self.state.global_step = 0`), since `self.state.global_step % self.num_iterations == 0` evaluates to `True`, we generate and score completions for the entire batch. Here we also store the old model's token probabilities (`old_per_token_logps`):

https://github.com/huggingface/trl/blob/d625c5533a6b1c84d3565c8080857f6bb81c538a/trl/trainer/grpo_trainer.py#L840-L848

3. Next, we compute the loss and update the policy ($\pi_{\theta}$). Note that `per_token_logps = old_per_token_logps` for now:

https://github.com/huggingface/trl/blob/d625c5533a6b1c84d3565c8080857f6bb81c538a/trl/trainer/grpo_trainer.py#L1043-L1065

5. However, on the next iteration (`global_step = 1`), we **do not** generate new completions. Instead, we reuse the inputs and completions from the previous iteration to perform a second policy update on the same batch. This step aligns with the intended GRPO iterations mentioned in Algorithm 1.

6. In this second iteration, `old_per_token_logps` remain as originally computed at `global_step = 0`, but now `per_token_logps` are computed using the updated policy, causing these probabilities to diverge.

7. On the subsequent iteration (`global_step = 2`), we generate a new batch of prompts and completions, restarting the cycle.

**Comment by YooSungHyun** [comment]
@LeonEricsson 

Damn! I completely misunderstood!!!
I totally missed this line:
`if self.state.global_step % self.num_iterations == 0 or buffered_inputs is None:`

Sorry for the trouble, and thank you so much for pointing it out!
